lib/puppet/util/windows/taskscheduler2.rb

```diff
@@ -0,0 +1,707 @@
+    # time = time + max_time[:minute].to_i * 60                           unless max_time[:minute].nil?
+    # time = time + max_time[:second].to_i                                unless max_time[:second].nil?
+
+    duration_hash_to_seconds(max_time) * 1000
+  end
+
+  # Sets the maximum length of time, in milliseconds, that the task can run
+  # before terminating. Returns the value you specified if successful.
+  #
+  def max_run_time=(max_run_time)
+    raise Error.new(_('No currently active task. ITask is NULL.')) if @pITask.nil?
+    raise TypeError unless max_run_time.is_a?(Numeric)
+
+    @pITask.SetMaxRunTime(max_run_time)
+    max_run_time
+  end
+
+  # Returns whether or not the scheduled task exists.
+  # DONE
+  def exists?(job_name)
+    # task name comparison is case insensitive
+    enum_task_names.any? { |name| name.casecmp(job_name) == 0 }
+  end
+
+  private
+
+  # Recursively converts a WIN32OLE Object in to a hash.  This method
+  # only outputs the Get Methods for an Object that has no parameters on the methods
+  # i.e. they are Object properties
+  #
+  def win32ole_to_hash(win32_obj)
+    hash = {}
+
+    win32_obj.ole_get_methods.each do |method|
+      # Only interested in get methods with no params i.e. object properties
+      if method.params.count == 0
+        value = nil
+        begin
+          value = win32_obj.invoke(method.name)
+        rescue WIN32OLERuntimeError => err
+          # E_NOTIMPL 0x80004001 from # https://msdn.microsoft.com/en-us/library/windows/desktop/aa378137(v=vs.85).aspx
+          if err.message =~ /80004001/m
+            # Somehow the interface has the OLE Method, but the underlying object does not implement the method.  In this case
+            # just return nil and swallow the error
+            value = nil
+          else
+            raise
+          end
+        end
+        if value.is_a?(WIN32OLE)
+          # Recurse into the object tree
+          hash[method.name.downcase] = win32ole_to_hash(value)
+        else
+          hash[method.name.downcase] = value
+        end
+      end
+    end
+
+    hash
+  end
+
+  # Recursively sets properties on a WIN32OLE Object from a hash.  This method
+  # only set the Put Methods for an Object
+  def set_properties_from_hash(ole_obj, prop_hash)
+    method_list = ole_obj.ole_put_methods.map { |method| method.name.downcase }
+
+    prop_hash.each do |k,v|
+      if v.is_a?(Hash)
+        set_properties_from_hash ole_obj.invoke(k), v
+      else
+        new_val = v
+        # Ruby 2.3.1 crashes when setting an empty string e.g. '', instead use nil
+        new_val = nil if v.is_a?(String) && v.empty?
+        ole_obj.setproperty(k,new_val) if method_list.include?(k.downcase)
+      end
+    end
+  end
+
+  def populate_hash_from_trigger(task_trigger)
+    return nil if task_trigger.nil?
+
+    hash = win32ole_to_hash(task_trigger)
+
+    hash['type_name'] = task_trigger.ole_type.name
```

Comment:
So `type_name` ends up being a hint for future deserialization / hydration of an appropriate COM trigger class name.... Looking at this further, I think we only need the trigger type enum value.